server/src/main/java/org/elasticsearch/common/lucene/search/function/MinScoreScorer.java

```diff
@@ -19,28 +19,25 @@
 
 package org.elasticsearch.common.lucene.search.function;
 
-import java.io.IOException;
-
 import org.apache.lucene.search.DocIdSetIterator;
-import org.apache.lucene.search.ScoreCachingWrappingScorer;
 import org.apache.lucene.search.Scorer;
 import org.apache.lucene.search.TwoPhaseIterator;
 import org.apache.lucene.search.Weight;
 
+import java.io.IOException;
+
 /** A {@link Scorer} that filters out documents that have a score that is
  *  lower than a configured constant. */
 final class MinScoreScorer extends Scorer {
 
     private final Scorer in;
     private final float minScore;
 
+    private int curDoc = -1;
+    private float curScore;
+
     MinScoreScorer(Weight weight, Scorer scorer, float minScore) {
         super(weight);
-        if (scorer instanceof ScoreCachingWrappingScorer == false) {
-            // when minScore is set, scores might be requested twice: once
-            // to verify the match, and once by the collector
-            scorer = new ScoreCachingWrappingScorer(scorer);
-        }
```

Comment:
ScoreCachingWrapperScorer is now a Scorable, so this no longer works.  Instead I've just duplicated the caching logic here.  Given that SCWS is only ever applied by Collectors, and this is a real Scorer (so won't be generated by a Collector), we don't need to worry about double wrapping.

Reply:
you probably need to fix the two-phase iterator to call `this.score()` rather than `in.score()` as well then, or otherwise it by-passes caching?

Reply:
actually maybe the two-phase iterator could just do `curScore = in.score(); return curScore >= minScore;` and then score() wouldn't even have to check the current doc ID

Reply:
++

I also changed MinScoreScorerTests to check that score() only gets called once on the wrapped Scorable.